Reimplement the progress display in wide-layout attachments

squelart opened the pull request:
#### 31c3e79e949e6fa5e661a2671baf8e8d23953c71
<pre>
Reimplement the progress display in wide-layout attachments
<a href="https://bugs.webkit.org/show_bug.cgi?id=255100">https://bugs.webkit.org/show_bug.cgi?id=255100</a>
rdar://problem/107716507

Reviewed by Aditya Keerthi.

When a wide-layout attachment has an attribute &quot;progress&quot; with a number, it is now directly handled by the top-level attachment element, by displaying a pie chart; previously it was handled by the inner attachment&apos;s legacy code.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::attachmentProgressIdentifier):
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
(WebCore::HTMLAttachmentElement::updateProgress):
(WebCore::HTMLAttachmentElement::parseAttribute):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-progress):

Canonical link: <a href="https://commits.webkit.org/262758@main">https://commits.webkit.org/262758@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/558cd0cbe3fa664c100e800b924ee37ddfd81c3f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2267 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2291 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2375 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3194 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2295 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2241 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2380 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2349 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2044 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2288 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/2015 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2036 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3052 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4 "21 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2017 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1911 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2098 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2025 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3211 "Passed tests") | 
| | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2063 "Passed tests") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1874 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2080 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2027 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/619 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2006 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2201 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -275,6 +293,29 @@ class AttachmentSaveEventListener final : public EventListener {
+    bool validProgress = false;
+    float value = progress.toFloat(&validProgress);
+    if (validProgress && std::isfinite(value)) {
+        StringBuilder style;
```
I think we prefer to use `makeString` / `makeAtomString` when possible.

squelart replied:
Thanks for the tip, much better indeed!

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -275,6 +293,29 @@ class AttachmentSaveEventListener final : public EventListener {
+        if (value < 0.0)
+            style.append("0"_s);
+        else if (value > 1.0)
+            style.append("1"_s);
```
Can we use `std::clamp` to normalize the value, and avoid these conditionals?

squelart replied:
I wanted to avoid doing a float->string conversion when it can be bypassed here (since we already have the string in most cases).
Would you still prefer clamping + to-string'ing the value?

pxlcoder replied:
We could clamp and only convert to string if the values are not equal.

I don't feel too strongly about this though.

squelart replied:
> only convert to string if the values are not equal

Possible, but I'm afraid it would become more complex code!
Thanks for your understanding.

pxlcoder commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -80,6 +80,25 @@ attachment#attachment-preview {
+}
+
+/* FIXME: Combine border into attachment-progress above, when rdar://107621207 is fixed. */
+div#attachment-progress-circle {
```
Would we also be able to remove this `div` entirely when rdar://107621207 is fixed? Other than that, it's not obvious to me why we need both.

squelart replied:
Oh, that's what I meant: Move the border styling to the above div, and remove this div. I'll rephrase for clarity...

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -54,6 +54,7 @@
 #include <wtf/IsoMallocInlines.h>
 #include <wtf/UUID.h>
 #include <wtf/URLParser.h>
+#include <wtf/text/StringBuilder.h>
```
This include can be removed now.

squelart committed:
Reimplement the progress display in wide-layout attachments
https://bugs.webkit.org/show_bug.cgi?id=255100
rdar://problem/107716507

Reviewed by Aditya Keerthi.

When a wide-layout attachment has an attribute "progress" with a number, it is now directly handled by the top-level attachment element, by displaying a pie chart; previously it was handled by the inner attachment's legacy code.

* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::attachmentProgressIdentifier):
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
(WebCore::HTMLAttachmentElement::updateProgress):
(WebCore::HTMLAttachmentElement::parseAttribute):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-progress):

Canonical link: https://commits.webkit.org/262758@main
